### PR TITLE
Upgrade notify to v5

### PR DESCRIPTION
**Pull request by alexkirsz**
This also switches from the default FSEvents implementation to the new kqueue-based implementation, which I've measured to be significantly faster.

However, the new implementation also runs into the `Too many open files (os error 24)` error much earlier. I had to add an `eprintln` log to surface this error, which would otherwise go unreported by turbopack. A solution right now is to manually run `ulimit -n unlimited` in the shell session you want to run turbopack in. This could possibly be done by turbopack itself instead.

### CODE REVIEW

File: crates/turbo-tasks-fs/src/lib.rs [hunk 1]
```diff
@@ -131,76 +138,35 @@ impl DiskFileSystem {
...
-                            if let Some(parent) = destination.parent() {
+            loop {
+                match rx.recv() {
+                    Ok(Ok(events)) => {
+                        for event in events {
+                            if let Some(parent) = event.path.parent() {
                                 batched_invalidate_path_dir.insert(PathBuf::from(parent));
                             }
+                            batched_invalidate_path_and_children.insert(event.path.clone());
+                            batched_invalidate_path_and_children_dir.insert(event.path);
```
Comment (alexkirsz) [inline crates/turbo-tasks-fs/src/lib.rs:157]: The new debouncer implementation has no granularity as to the type of event. `notify` itself now only reports "raw" events. So I'm now doing the same thing for all event types, but I'm not sure whether that's correct.

File: crates/turbo-tasks-fs/src/lib.rs [hunk 2]
```diff
@@ -112,15 +113,21 @@ impl DiskFileSystem {
         }
         let invalidator_map = self.invalidator_map.clone();
         let dir_invalidator_map = self.dir_invalidator_map.clone();
-        let root = self.root.clone();
         // Create a channel to receive the events.
-        let (tx, rx) = channel();
+        let (tx, rx) = channel::<Result<Vec<DebouncedEvent>, Vec<notify::Error>>>();
```
Comment (alexkirsz) [inline crates/turbo-tasks-fs/src/lib.rs:117]: We could use a tokio channel or a crossbeam channel here instead.

File: crates/turbo-tasks-fs/src/lib.rs (hunk 1 above)
Comment (sokra) [reply crates/turbo-tasks-fs/src/lib.rs:157]: How does these raw events look like?

Comment (sokra) [reply crates/turbo-tasks-fs/src/lib.rs:157]: These raw events look promising...

Comment (sokra) [reply crates/turbo-tasks-fs/src/lib.rs:157]: But debouncer seems fine too... Let's keep that until we run into problems...

**Review by jridgewell — APPROVED** [review]

**Comment by wbinnssmith** [comment]
> This also switches from the default FSEvents implementation to the new inotify-based implementation, which I've measured to be significantly faster.

Nit: do you mean kqueue here? I think inotify is a Linux thing, whereas FSEvents is on macOS

**Comment by alexkirsz** [comment]
@wbinnssmith Yes, I meant kqueue.

File: crates/turbo-tasks-fs/Cargo.toml [hunk 3]
```diff
@@ -23,7 +23,8 @@ include_dir = { version = "0.7.2", features = ["nightly"] }
 jsonc-parser = { version = "0.21.0", features = ["serde"] }
 lazy_static = "1.4.0"
 mime = "0.3.16"
-notify = "4.0.17"
+notify = { version = "5.0.0", features = ["macos_kqueue"] }
```
Comment (sokra) [inline crates/turbo-tasks-fs/Cargo.toml:26]: Could we ship the PR without kqueue? notify v5 is probably still a good change

**Comment by ForsakenHarmony** [comment]
Have also looked into this a while ago, don't know if it's of any value, but here's the old patch
https://gist.github.com/ForsakenHarmony/80b2d574670158dc0d7cc1b0930b504a

File: crates/turbo-tasks-fs/Cargo.toml (hunk 3 above)
Comment (ForsakenHarmony) [reply crates/turbo-tasks-fs/Cargo.toml:26]: Is there a downside?

**Pull request closed without merging**